### ISSUE TITLE
[FEATURE REQUEST]: Add threshold to SEBO to prevent over-exploration of low sparsity inputs.

### ISSUE THREAD

**Issue by CompRhys** [facebook/Ax#3518]
### Motivation

Is there interest in upstreaming some changes that add a threshold value below which further decreases in L-p Norms are not useful for SEBO?

For example if you have a problem where selecting less than N channels has no benefit over selecting N then it makes sense to threshold at N.

### Describe the solution you'd like to see implemented in Ax.

Changes are fairly simple `relu(L_norm - threshold)+threshold` and I can make PRs. There is a question of how widespread the addition should be, should all the norms in `botorch/acquisition/penalized.py` be updated? Whilst not needed for my use case should a `max_threshold` also be added for the completeness of the API?

### Code of Conduct

- [x] I agree to follow Ax's Code of Conduct

**Comment by Balandat** [comment]
I think this _could_ potentially be useful - curious if @dme65 has come across this before. I think if we implemented that capability we should add it to the API as well. Saying that, not sure exactly what you mean by that - anything more than defining some option key and grabbing that in the constructor? 

https://github.com/facebook/Ax/blob/d6fe6dbd618d1c78d20d0ca84c4732113f54e019/ax/models/torch/botorch_modular/sebo.py#L63

**Comment by CompRhys** [comment]
yes exactly in local solution I have 
```
        self.sparsity_lower_bound: Tensor = options.pop(
            "sparsity_lower_bound", torch.tensor(0.0)
        )
```

and then updated the penalty functions to include:

```
def L1_norm_func(
    X: Tensor, init_point: Tensor, lower_bound: Tensor = torch.tensor(0.0)
) -> Tensor:
    r"""L1_norm takes in a a `batch_shape x n x d`-dim input tensor `X`
    to a `batch_shape x n x 1`-dimensional L1 norm tensor. To be used
    for constructing a GenericDeterministicModel.
    """
    return (
        torch.relu(
            torch.linalg.norm((X - init_point), ord=1, dim=-1, keepdim=True)
            - lower_bound
        )
        + lower_bound
    )
```

The L0 penalty function is inside Botorch but I needed to edit that also.

**Comment by CompRhys** [comment]
Shall I make a PR for this? Can either do it touching BoTorch or change SEBO such that it doesn't reuse the L0 code in botorch.

**Comment by dme65** [comment]
@CompRhys 

Thank you for the suggestion!

> Is there interest in upstreaming some changes that add a threshold value below which further decreases in L-p Norms are not useful for SEBO? For example if you have a problem where selecting less than N channels has no benefit over selecting N then it makes sense to threshold at N.

I'm not a fan of this idea as we are treating sparsity as a separate objective in SEBO and do multi-objective optimization, so the suggestion here corresponds to constraining the sparsity objective in the direction of optimization (we don't support constraints on objectives with the beyond objective thresholds in Ax). Wouldn't it be more natural to use a stricter objective threshold for the other metric(s) to avoid exploring regions with low sparsity if that isn't desired?

> Changes are fairly simple relu(L_norm - threshold)+threshold and I can make PRs. 

This likely won't work well since the gradient of the sparsity term will be zero when `L_norm < threshold`. This may cause issues when optimizing the acquisition function.

> Whilst not needed for my use case should a max_threshold also be added for the completeness of the API?

This is already supported by passing down `sparsity_threshold` to `SEBOAcquisition` (corresponds to the objective threshold for sparsity).

**Comment by CompRhys** [comment]
> I'm not a fan of this idea as we are treating sparsity as a separate objective in SEBO and do multi-objective optimization, so the suggestion here corresponds to constraining the sparsity objective in the direction of optimization (we don't support constraints on objectives with the beyond objective thresholds in Ax). Wouldn't it be more natural to use a stricter objective threshold for the other metric(s) to avoid exploring regions with low sparsity if that isn't desired?

The desire here is scientific in that a lot of experiments have already been done in the very low sparsity systems by others. Whilst we don't have compatible data such that we can include it in training there's negligible real value in running those experiments just to try get the optimizer to stop suggesting them. Nothing in the literature satisfies the requirements but there's also no underlying reason why these very low sparsity systems couldn't.

Not sure of whether there's a more subtle difference i'm missing between outcome constraints and objective constraints here, but if we instead had a GP surrogate model to predict the Lp norm we could put a constraint on that inside Ax and use EHVI type methods (it's actually what the examples show: https://ax.dev/docs/tutorials/gpei_hartmann_loop/). I'm skeptical that we could learn a good L0 surrogate but within SEBO the implementation of the algorithm means that we don't need to learn it as we're adding a differentiable model that directly predicts it to the model list. 

> This likely won't work well since the gradient of the sparsity term will be zero when `L_norm < threshold`. This may cause issues when optimizing the acquisition function.

I will need to look into this more, I had previously used GA to optimize the sparsity within a MOO before encountering SEBO so hadn't given any thought to the gradient flow. My interest in SEBO is that it's significantly faster than the GA to generate candidates.

> This is already supported by passing down `sparsity_threshold` to `SEBOAcquisition` (corresponds to the objective threshold for sparsity).

Yes I saw this when digging into the code more to try make minimal PR in #3564.

**Comment by lena-kashtelyan** [comment]
Centralized SEBO issues here: https://github.com/facebook/Ax/issues/3828
